hr_timesheet_sheet/models/hr_timesheet_sheet.py

```diff
@@ -0,0 +1,628 @@
+                    ('project_id', '=', row[0].project_id.id),
+                    ('task_id', '=', task),
+                    ('date', '<=', self.date_end),
+                    ('date', '>=', self.date_start),
+                    ('employee_id', '=', self.employee_id.id),
+                    ('sheet_id', '=', self.id),
+                    ('company_id', '=', self.company_id.id),
+                ])
+                if allow_empty_rows and self.add_line_project_id:
+                    check = any([l.unit_amount for l in row])
+                else:
+                    check = not all([l.unit_amount for l in row])
+                if check:
+                    ts_row.filtered(
+                        lambda t: t.name == '/' and not t.unit_amount).unlink()
+
+    # ------------------------------------------------
+    # OpenChatter methods and notifications
+    # ------------------------------------------------
+
+    @api.multi
+    def _track_subtype(self, init_values):
+        if self:
+            record = self[0]
+            if 'state' in init_values and record.state == 'confirm':
+                return 'hr_timesheet_sheet.mt_timesheet_confirmed'
+            elif 'state' in init_values and record.state == 'done':
+                return 'hr_timesheet_sheet.mt_timesheet_approved'
+        return super(Sheet, self)._track_subtype(init_values)
+
+    @api.model
+    def _needaction_domain_get(self):
+        empids = self.env['hr.employee'].search(
+            [('parent_id.user_id', '=', self.env.uid)])
+        if not empids:
+            return False
+        return ['&', ('state', '=', 'confirm'),
+                ('employee_id', 'in', empids.ids)]
+
+
+class SheetLine(models.TransientModel):
+    _name = 'hr_timesheet.sheet.line'
+    _description = 'Timesheet Sheet Line'
+
+    sheet_id = fields.Many2one(
+        comodel_name='hr_timesheet.sheet',
+        ondelete='cascade',
+    )
+    date = fields.Date(
+        string='Date',
+    )
+    project_id = fields.Many2one(
+        comodel_name='project.project',
+        string='Project',
+    )
+    task_id = fields.Many2one(
+        comodel_name='project.task',
+        string='Task',
+    )
+    value_x = fields.Char(
+        string='Date Name',
+    )
+    value_y = fields.Char(
+        string='Project Name',
+    )
+    unit_amount = fields.Float(
+        string="Quantity",
+        default=0.0,
+    )
+    count_timesheets = fields.Integer(
+        default=0,
+    )
+
+    @api.onchange('unit_amount')
+    def onchange_unit_amount(self):
+        """This method is called when filling a cell of the matrix.
+        It checks if there exists timesheets associated  to that cell.
+        If yes, it does several comparisons to see if the unit_amount of
+        the timesheets should be updated accordingly."""
+        self.ensure_one()
+        if self.unit_amount < 0.0:
+            self.write({'unit_amount': 0.0})
```

Comment:
Sorry to insist, but I don't think this module should prevent the user to enter negative amounts.
The above two lines should be removed IMO.

Reply:
It's not that simply to just remove this two lines. The <code>onchange_unit_amount</code> method should be fully rewritten to work with negative hours without doing strange things.

But finally, I won't change that because the intention of this PR is not just doing an "exact" copy of the v10, but it also tries to include some improvements. And one of this improvements is not to allow negative hours, because to me unit_amount is an interval and intervals, from a mathematical point of view, are never negative. To me, working with negative hours is a wrong usability, and it potentially can add obscure bugs. If hypothetically exists an user that does that, it is 1% of all the users. And for just that 1% I won't change that behavior. This 1% will have to make their adapted customization.

Reply:
Thanks for the explanation. I agree with your points.